src/main/java/com/sunnylabs/tracegenerator/Configuration.java

```diff
@@ -0,0 +1,282 @@
+                Operation o = getRandom(ops);
+                available.remove(o);
+                o.addCall(getRandom(available));
+            }
+        }));
+
+        // cross service calls per app
+        apps.values().forEach(a -> {
+            Collection<Service> services = a.getServices().values();
+            List<Service> available = new ArrayList<>(services);
+            while (available.size() > 1) {
+                Service s = getRandom(services);
+                Operation o = getRandom(s.getOperations());
+
+                available.remove(s);
+                Operation op2 = getRandom(getRandom(available).getOperations());
+
+                o.addCall(op2);
+            }
+        });
+
+        // cross app calls
+        for (int i = 0; i < apps.size(); i++) {
+            Operation random = getRandom(getRandom(getRandom(apps).getServices()).getOperations());
+            Operation target = getRandom(getRandom(getRandom(apps).getServices()).getOperations());
+            while (target.getApplication().equals(random.getApplication())) {
+                target = getRandom(getRandom(getRandom(apps).getServices()).getOperations());
+            }
+            random.addCall(target);
+        }
+    }
+
+    private <T> T getRandom(Map<?, T> from) {
+        return new ArrayList<>(from.values()).get(new Random().nextInt(from.size()));
+    }
+
+    private <T> T getRandom(Collection<T> from) {
+        return new ArrayList<>(from).get(new Random().nextInt(from.size()));
+    }
+
+
+    private Application createRandomApp(List<String> serviceNames, List<String> operationNames) {
+        Application a = new Application();
+        Map<String, Service> services = new HashMap<>();
+        int desiredServices = Math.min(servicesPerApp, serviceNames.size());
+        for (int i = 0; i < desiredServices; i++) {
+            int idx = new Random().nextInt(serviceNames.size());
+            String name = serviceNames.get(idx);
+            serviceNames.remove(idx);
+            Service service = createRandomService(new ArrayList<>(operationNames));
+            services.put(name, service);
+        }
+        a.setServices(services);
+        return a;
+    }
+
+    private Service createRandomService(List<String> operationNames) {
+        Service s = new Service();
+        List<Operation> operations = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            int idx = new Random().nextInt(operationNames.size());
+            String name = operationNames.get(idx);
+            operationNames.remove(idx);
+            operations.add(new Operation(name));
+        }
+        s.setOperations(operations);
+        return s;
+    }
+
+    private void fixOperationReferences(List<Operation> calls) {
+        for (int i = 0; i < calls.size(); i++) {
+            final int index = i;
+            existingOperation(calls.get(index)).ifPresent(e -> calls.set(index, e));
```

Comment:
should the operation be removed from the list if not present?

Reply:
No, because then it's a stub reference (it's valid to have operations called that aren't defined elsewhere, following a principle of minimizing required config). This method replaces any stub operations that are defined elsewhere with references to those operations.